### ISSUE TITLE
End-to-end tests

### ISSUE THREAD

**Issue by williamboman** [thelounge/thelounge#2359]
I think having end-to-end tests would greatly increase the confidence when landing PRs and reduce the need for a lot of manual testing. I've been interested in trying out https://www.cypress.io/ for some time and I'd like to use The Lounge as the testing target if there's an interest in having these.

**Comment by McInkay** [comment]
Yeah, totally happy to someone to jump on this.

I would say that if the tests end up being flaky (nothing against your coding, it's just every experience of end to end tests that I've found have been very flaky) then I'd rather they weren't run on the build on master, because we'll just end up having to rerun. It's fine to have it on PRs but not on master if flaky.

**Comment by astorije** [comment]
I'd love to see this too. Cypress is what's on everyone's mind these days, so definitely worth a shot! Thanks @williamboman!

@McInkay, let's iron out the details when there is something ready. I'm open to any tweak for now, just want to see where it leads us.

**Comment by williamboman** [comment]
I have my fair share of experience with e2e tests and I share your concerns @McInkay 😅. As to setting up the back-end I don't have any clear ideas. For starters I'll just run against Freenode. Any ideas/pointers regarding networks/channels/users/mocking/etc. would be appreciated.

**Comment by astorije** [comment]
You might get yourself k-lined or banned or whatnot, so I suggest you first either run a dummy IRC server locally (don't worry about CI for now imo) or run against one of those open servers that are meant for testing (InspIRCd?).

@McInkay had ideas to make TL shippable with a built-in IRC server, which could help here, and until that happens, we could set up a simple IRCd in CI.

**Comment by McInkay** [comment]
Yeah I'm working on a small IRCd that will support most of the features the lounge does, and can be added as a plugin. https://github.com/mcinkay/ircd Feel free to give it a star to keep an eye on it.

It's quite early, so it doesn't support everything we need, but if you are going to add the tests slowly, I might be able to add features at the same rate as you need them. I'm going to be doing some work on it this week.

If it isn't far enough along when you start, then you could use another one but all the other node ones are reaaaaaally out of date and horrible to try and get running. I agree with @astorije ... don't use a hosted one but run one in the script.

**Comment by astorije** [comment]
Hey @williamboman, have you had any success on this? :)

**Comment by williamboman** [comment]
> Hey @williamboman, have you had any success on this? :)

Unfortunately not anything so far. I will most likely not have time to look into this until after the summer. If someone wants to pick up the ball, please do!
